packages/web/components/related-assets/related-assets.tsx

```diff
@@ -1,55 +1,153 @@
+import { ObservableQueryPool } from "@osmosis-labs/stores";
+import { observer } from "mobx-react-lite";
 import Image from "next/image";
+import Link from "next/link";
+import { FunctionComponent } from "react";
 import { useTranslation } from "react-multi-lang";
 
+import { useAssetInfoConfig } from "~/hooks";
+import { useStore } from "~/stores";
+import { CoinBalance } from "~/stores/assets";
+
 import caretDown from "../../public/icons/caret-down.svg";
 
-export default function RelatedAssets() {
+const numberOfAssetsToDisplay = 8;
+
+const RelatedAssets: FunctionComponent<{
+  memoedPools: ObservableQueryPool[];
+  tokenDenom: string;
+}> = observer(({ memoedPools, tokenDenom }) => {
   const t = useTranslation();
-  return (
+
+  const { assetsStore, queriesExternalStore } = useStore();
+
+  const balances = assetsStore.nativeBalances.concat(
+    assetsStore.unverifiedIbcBalances
+  );
+
+  const assetsToDisplay = new Set<string>();
+  for (const pool of memoedPools) {
```

Comment:
Have you checked whether doing these operations inline and not inside a memo, doesn't generate too many re-renders?

Reply:
All those variables are only read except for `assetsToDisplay` which is a plain local variable. Not a state or mobx reactive variable. No line in this component should cause a rerender. Maybe memoizing the first part could lead to a slight performance increase but is negligible as all input props and reactive variables are used in this calculation